### PR TITLE
INT-4174: Expose Channel Subscriber Count

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-4174

I considered just using the channel counter but I suppose there is a (small) possibility
that someone might possibly use a custom dispatcher and subscribe to it directly.

__cherry-pick to 4.3.x__

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractSubscribableChannel.java [hunk 1]
```diff
@@ -37,19 +38,30 @@
...
 
 	private final AtomicInteger handlerCounter = new AtomicInteger();
 
+	@Override
+	public int getSubscriberCount() {
+		MessageDispatcher dispatcher = getRequiredDispatcher();
+		if (dispatcher instanceof AbstractDispatcher) {
+			return ((AbstractDispatcher) dispatcher).getHandlerCount();
+		}
+		return this.handlerCounter.get();
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractSubscribableChannel.java]: Good idea - working on it now...

Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractSubscribableChannel.java]: Great! 
But I think we can't backport it to `4.3.x`. 
Therefore that work should be done in the separate PR.

What is in your mind ?

Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractSubscribableChannel.java]: Just cherry-pick the first commit.

Let me know if you prefer a separate PR.

**Review by artembilan — APPROVED** [review]
Merging and cherry-picking to `4.3.x` only first commit...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/96ff547c8c06c9e465270485890974ccfbdfdebb and cherry-picked to `4.3.x` only first commit.

Also note the fix for JPA tests race condition.
